### PR TITLE
Expression evaluator fixes

### PR DESCRIPTION
**Pull request by ZauberNerd**
This PR contains a few fixes and improvements for the new expression evaluator:
* Adds more debug output about which expression is going to be evaluated
* Use exported `IsTruthy()` function from the `exprparser` package inside `EvalBool()`
* Fail act when an expression contains errors and cannot be evaluated

Fixes: #1008 
Ref: #971

### CODE REVIEW

**Comment by ZauberNerd** [comment]
@ChristopherHX do you remember why you added this debug log in #971 https://github.com/nektos/act/commit/7dbf3fcb96e2d70b3436b718c1f4a9fe4b169a4f#diff-fa62a470d83900f015861d603c59c5e371354fb0cdf36089a274d40ee409029eR139?
Since the error is being returned it will be printed eventually.

**Comment by ChristopherHX** [comment]
I restored the function 1:1 as it was prior running steps after job failure

https://github.com/nektos/act/blob/v0.2.25/pkg/common/executor.go#L139

I'm not the original author of this debug print
> Since the error is being returned it will be printed eventually.

Yes it will **eventually** get printed. I think it good for debugging to always print any error of the then executor.

**Comment by ZauberNerd** [comment]
> Yes it will eventually get printed. I think it good for debugging to always print any error of the then executor.

In my opinion debug logging should be used to provide more information that otherwise isn't visible in order to be able to understand what is happening.
In this case the error is already being printed, so logging the error into the debug log doesn't make much sense.

Below is the test run of https://github.com/nektos/act/pull/971/files#diff-4526efc54b2af950bff29a71c13adadb9f22132374291f264c843cc3cfd51d21

```
$ ./dist/local/act -W ./pkg/runner/testdata/non-existent-action/push.yml 
[non-existent-action/nopanic] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[non-existent-action/nopanic]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform= username= forcePull=false
[non-existent-action/nopanic]   🐳  docker create image=ghcr.io/catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[non-existent-action/nopanic]   🐳  docker run image=ghcr.io/catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[non-existent-action/nopanic]   🐳  docker exec cmd=[mkdir -m 0777 -p /var/run/act] user=root workdir=
[non-existent-action/nopanic] ⭐  Run ./path/to/non-existent-action
[non-existent-action/nopanic]   ❌  Failure - ./path/to/non-existent-action
[non-existent-action/nopanic] file does not exist
Error: Job 'nopanic' failed
```

**Review by KnisterPeter — DISMISSED** [review]

**Comment by catthehacker** [comment]
> In my opinion debug logging should be used to provide more information that otherwise isn't visible in order to be able to understand what is happening. In this case the error is already being printed, so logging the error into the debug log doesn't make much sense.

you could make it a trace instead debug

**Review by EldarKurbanov — APPROVED** [review]

**Review by mvrshvl — APPROVED** [review]

**Review by mvrshvl — APPROVED** [review]

**Comment by ZauberNerd** [comment]
@ChristopherHX @catthehacker so what's the status here? Should I:
* restore the old behaviour and log the error as debug?
* log the error as trace?
* leave it as is?

In case you'd like to keep a log in the `Then` executor, I'd probably change the log to include a prefix, such as `Error in Then executor: %+v`` or similar, because @KnisterPeter and me had a hard time figuring out where the log came from.

**Comment by catthehacker** [comment]
I don't really have any preference, I'll leave it up to you
Just wanted to give another option due to https://github.com/nektos/act/pull/928

File: pkg/runner/expression.go [hunk 1]
```diff
@@ -262,7 +266,7 @@ func EvalBool(evaluator ExpressionEvaluator, expr string) (bool, error) {
 			result = t != 0
 		}
 	default:
-		return false, fmt.Errorf("Unable to map return type to boolean for '%s'", expr)
+		return false, fmt.Errorf("Unable to map type '%t' to boolean for '%s'", evaluated, expr)
```
Comment (ChristopherHX) [inline pkg/runner/expression.go:269]: I thought the default case is impossible to reach for a conforming evaluator.
Any Object != null => true, except empty string
- `if: github.event.pull_request` ( true if PR )
- `if: fromJSON('{}')` ( always true )

<Details>
<Summary>This weird workflow only fails in act, the last step is expected to run</Summary>

```yaml
on:
- push
- pull_request
jobs:
  m:
    runs-on:
    - ubuntu-latest
    steps:
    - run: |
                echo PR!
      shell: bash
      if: |
                github.event.pull_request
    - run: |
                echo Push!
      shell: bash
      if: |
                !github.event.pull_request
    - run: |
                echo empty obj!
      shell: bash
      if: |
                fromJSON('{}')
```

</Details>

**Review by ChristopherHX — APPROVED** [review]
Works: I'm ok with removing the debug print. You won't need to address my review comment, which is just an example ( and a bug in act ) why I prefer to use actions/runner locally.

In a later change, we might want to connect this error to an actual job.
```
Error:   ❌  Error in if-expression: "if: fromJSON('{}')||github.event.pull_request || !github.event.pull_request
" (Unable to map type 'map[]' to boolean for 'fromJSON('{}')||github.event.pull_request || !github.event.pull_request
')
```

File: pkg/runner/expression.go (hunk 1 above)
Comment (ZauberNerd) [reply pkg/runner/expression.go:269]: Good catch, I'll update the function to handle maps.

**Review by KnisterPeter — APPROVED** [review]
Nice

**Comment by cplee** [comment]
@ZauberNerd - is this a breaking change? looks like the tests had to be updated from `"` -> `'`...any concern with that having impact on users?

**Comment by ZauberNerd** [comment]
> @ZauberNerd - is this a breaking change? looks like the tests had to be updated from `"` -> `'`...any concern with that having impact on users?

These workflows are invalid according to GitHub and strings can only use single quotes: https://docs.github.com/en/actions/learn-github-actions/expressions#literals

Here's a workflow run of the "issue-597" workflow: https://github.com/ZauberNerd/act-test/actions/runs/1889956547
```
[Invalid workflow file: .github/workflows/test.yml#L17](https://github.com/ZauberNerd/act-test/actions/runs/1889956547/workflow)
The workflow is not valid. .github/workflows/test.yml (Line: 17, Col: 13): Unexpected symbol: '"endsWith'. Located at position 1 within expression: "endsWith('Hello world', 'ld')"
```

Also, it doesn't look like they worked on master either, just the error while parsing the `if` condition didn't lead to a non-zero exit-code:
```
$ ./dist/local/act -W ./pkg/runner/testdata/issue-597/spelling.yaml 
[issue-597/my_first_job] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[issue-597/my_first_job]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform= username= forcePull=false
[issue-597/my_first_job]   🐳  docker create image=ghcr.io/catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[issue-597/my_first_job]   🐳  docker run image=ghcr.io/catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[issue-597/my_first_job]   🐳  docker exec cmd=[mkdir -m 0777 -p /var/run/act] user=root workdir=
[issue-597/my_first_job]   ❌  Error in if: expression - My first true step
```

With these changes it looks like this:
```
$ ~/act/dist/local/act -W ./issue-597/spelling.yaml 
[issue-597/my_first_job] 🚀  Start image=ghcr.io/catthehacker/ubuntu:act-latest
[issue-597/my_first_job]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:act-latest platform= username= forcePull=false
[issue-597/my_first_job]   🐳  docker create image=ghcr.io/catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[issue-597/my_first_job]   🐳  docker run image=ghcr.io/catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[issue-597/my_first_job]   🐳  docker exec cmd=[mkdir -m 0777 -p /var/run/act] user=root workdir=
[issue-597/my_first_job] ⭐  Run My first true step
[issue-597/my_first_job]   ☁  git clone 'https://github.com/actions/hello-world-javascript-action' # ref=main
[issue-597/my_first_job]   🐳  docker cp src=/home/ubuntu/.cache/act/actions-hello-world-javascript-action@main/ dst=/var/run/act/actions/actions-hello-world-javascript-action@main/
[issue-597/my_first_job]   🐳  docker exec cmd=[mkdir -p /var/run/act/actions/actions-hello-world-javascript-action@main/] user= workdir=
[issue-597/my_first_job]   🐳  docker exec cmd=[node /var/run/act/actions/actions-hello-world-javascript-action@main/index.js] user= workdir=
| Hello Renst the Octocat!
[issue-597/my_first_job]   ⚙  ::set-output:: time=22:26:32 GMT+0000 (GMT)
| The event payload: {}
[issue-597/my_first_job]   ✅  Success - My first true step
```

**Review by ChristopherHX — APPROVED** [review]
I also wouldn't classify the test workflow corrections as a breaking change, since it never used to work on github. I even blindly duplicated the invalid if in the composite test, in the past assuming these were valid.
_I would say deleting container on workflow failure was a breaking change introduced after 0.2.25, no tests no ci failure._

**Review by cplee — APPROVED** [review]

**Comment by cplee** [comment]
Thanks @ZauberNerd  - solid answer :)

**Pull request merged**